controller_manager/src/controller_manager.cpp

```diff
@@ -1151,8 +1151,9 @@ ControllerManager::RTControllerListWrapper::update_and_get_used_by_rt_list()
 std::vector<ControllerSpec> & ControllerManager::RTControllerListWrapper::get_unused_list(
   const std::lock_guard<std::recursive_mutex> &)
 {
-  assert(controllers_lock_.try_lock());
-  controllers_lock_.unlock();
+  {
+    std::lock_guard<std::recursive_mutex> guard(controllers_lock_);
```

Comment:
Nice catch, I actually prefer this type of scoped locks to manually operating them, much cleaner.

May I ask if you think we should keep the scope at all? 
@Karsten1987 can you chime in on this?

Reply:
Sorry, I'm not sure what you mean by "keep the scope"? If you mean if creating the new scope is necessary, I don't think it is, but I chose to create a new scope to mirror the old behavior of immediately locking and unlocking as closely as possible.

Reply:
Yeah I'm wondering if we could remove the scope around the lock guard but let's wait for @Karsten1987 on this

Reply:
I am not really familiar with this piece of code. I never really understood the business with used and unused lists. I believe @v-lopez is the right guy to answer here.

Reply:
The point was that this function takes a `lock_guard` as a parameter, to "force you" to own the mutex before calling it.

But, there's no way to check that the given `lock_guard` owns this particular mutex. So you try to lock it to ensure you own it.

It's done with try_lock and assert, because if you cannot lock, you made an error when writing the caller code and it should abort right away, it should not wait until someone else unlocks it and you lock, triggering a potential deadlock.

A better implementation would be to change the input argument, from `lock_guard` to `unique_lock`. Then replace the try_lock assert or any other lock with:
```
assert(lock.owns_lock());
assert(lock.mutex() == &controllers_lock_);
```

Reply:
I see! Thanks for explaining. Would it be okay to still keep this check in an `assert` since that means that they will not be run in release mode?

Reply:
so what do you recommend @v-lopez , given that we can't change the API to use unique_lock? How about the same mechanism with try_lock but throwing instead of the assert? That should maintain the current behaviour

Reply:
Agreed

Reply:
@Ace314159 could you please make these changes?

Reply:
@bmagyar Done

Reply:
Thanks!